ext/bg/js/text-source-map.js

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (C) 2020  Alex Yatskov <alex@foosoft.net>
+ * Author: Alex Yatskov <alex@foosoft.net>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+class TextSourceMap {
+    constructor(source, mapping=null) {
+        this._source = source;
+        this._mapping = (Array.isArray(mapping) ? TextSourceMap._normalizeMapping(mapping) : null);
```

Comment:
Could be simpler to make this `TextSourceMap._createMapping(source)` instead of `null` and add a boolean flag indicating if the mapping has been touched. Now it's initialized in multiple places, even for another `TextSourceMap`. There's a minor performance hit, though.

Reply:
The reason why I left it as null is for that performance reason. In most of the general use cases, the source map won't be modified, so deferring creation of the array should save allocation time and memory at the cost of only some `null` checks later on. `equals` also has a few additional `null` checks, but this function is only used for unit tests currently.

Reply:
I guess it's justified because dictionary lookup is such a frequent operation. It's better that the complexity is encapsulated in a class instead of having it in other places of code. I didn't realize that `equals` was for testing only.